### PR TITLE
Native query default parameter values: use ParameterValuePicker for single strings, numbers, and dates

### PR DESCRIPTION
**Pull request by oleggromov**
Part of [[Epic] Better UI/UX for picking default parameter values](https://github.com/metabase/metabase/issues/39211)

## What's done
We start from updating Default value picker in Native queries.

This PR does the following:

- 🐞 fixes: when we reset widget type (think `date/relative` -> `date/single`), reset the default value too
- 🐞 fixes ugly year picker lacking gap with the quarter selector in Quarter and Year picker
- disables Filter widget type when it has a single value (ID or None)
- uses Mantine for a single number (when you selected Number) or string (String + Input box)
- uses Mantine for date picker popovers
- stops using moment for generating ISO-style dates 2024-03-06
- tidies up `template-tags` utils

### How is it done
Instead of changing the original `ParameterValueWidget`, I have built a `ParameterValuePicker` component which reuses the former internally. 

```
tag, parameter -> ParameterValuePicker -> can I render it? ---YES--> do it
                                                          \---NO---> defer to ParameterValueWidget
```

With this approach, we can:
- precisely control what is rendered by the new component (it defaults to `false`)
- selectively roll out the change by choosing which components use `ParameterValuePicker`
- in case something breaks, easily revert the changes
- once we're done with Native queries, reuse the component for Dashboard parameters

### Component design
I deliberately made the component sub-tree (until it reaches `Date*` widgets) stateless. This way we don't struggle with the mismatch between the internal component state and what's coming to it from the store.

This was previously controlled by `commitImmediately` but as of this PR, I don't see why we would need internal state, which makes internals more complicated.

## Why it is done
- we still need to transition to Mantine and match UIs 
- there are multiple bugs and inadequacies in the current default value picking flow

## Screencast
Everything except shown in the screencast remains unchanged.

https://github.com/metabase/metabase/assets/2196347/9e9a9ba4-25ea-4310-8a3e-a8ca74ce954b

## Tests
- reset default value when changing widget type: covered by `sql-field-filter-[date|number|string].cy.spec.js`
- disabled widget type select: covered by `sql-field-filter.cy.spec.js`,  `31606.cy.spec.js`, `TagEditorParam.unit.spec.tsx`

### CODE REVIEW

File: frontend/src/metabase/parameters/components/ParameterValuePicker/OwnDatePicker.tsx [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+  const { value, parameter, onValueChange } = props;
+  const [isOpen, setIsOpen] = useState(false);
+  const formatted = formatParameterValue(value, parameter);
+  const parameterType = parameter.type as DateParameterType; // TODO fix types in Parameter
+
+  const openPopover = () => setIsOpen(true);
+  const closePopover = () => setIsOpen(false);
+
+  const [triggerRef, setTriggerRef] = useState<HTMLDivElement | null>(null);
+  const dropdownRef = useClickOutside(closePopover, null, [triggerRef]);
```
Comment (ranquild) [inline frontend/src/metabase/parameters/components/ParameterValuePicker/OwnDatePicker.tsx:36]: It should work out of the box, you shouldn't need this. Try `<Popover ... onChange={setIsOpen}>`

File: frontend/src/metabase/parameters/components/ParameterValuePicker/OwnDatePicker.tsx [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+    if (parameterType === "date/range") {
+      const now = getIsoDate();
+      return `${now}~${now}`;
+    }
+  }
+
+  return value;
+}
+
+function getIsoDate() {
```
Comment (ranquild) [inline frontend/src/metabase/parameters/components/ParameterValuePicker/OwnDatePicker.tsx:151]: Please use `dayjs`. This is our "modern" alternative to moment

File: frontend/src/metabase/parameters/components/ParameterValuePicker/OwnDatePicker.tsx [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+  Parameter,
+  ParameterType,
+} from "metabase-types/api";
+
+import {
+  TextInputIcon,
+  TextInputTrirgger,
+} from "./ParameterValuePicker.styled";
+
+export function OwnDatePicker(props: {
```
Comment (ranquild) [inline frontend/src/metabase/parameters/components/ParameterValuePicker/OwnDatePicker.tsx:22]: nit: In some newer files (but not all) we put nested components in subfolders (take a look at `metabase/querying`). We had an RFC about this and it was accepted, but I don't think that all people follow this approach

**Review by ranquild — APPROVED** [review]
It would be great to do small cleanup but this is a good first step

**Comment by oleggromov** [comment]
Thanks @ranquild, I'll address your suggestions in the next PR and merge this as is as we agreed.

**Pull request merged**